lib/spike.rb

```diff
@@ -0,0 +1,361 @@
+        api_call,
+      ].reject(&:empty?).join("\n        ")
+    end
+
+    def subresource_method_implementation
+      url_param = definition.params.find {|param| param.name.end_with?("_url")}
+      the_resource = resource.objects.first.singularize
+      s = "#{the_resource} = get(#{url_param.name}, accept: options[:accept])"
+      unless option_overrides.empty?
+        s << "\n        "
+        s << option_overrides.join("\n        ")
+      end
+      s << "\n        #{definition.verb.downcase}(#{the_resource}.rels[:#{resource.objects.last}].href, options)"
+    end
+
+    def option_overrides
+      definition.params.select(&:required).reject do |param|
+        definition.path.include?(":#{param.name}") || param.name.end_with?("_url")
+      end.map do |param|
+        normalization = ""
+        if !!param.enum
+          normalization = ".to_s.downcase"
+        end
+        "options[:#{param.name}] = #{param.name}#{normalization}"
+      end
+    end
+
+    def api_call
+      "#{definition.verb.downcase}(\"#{api_path}\", options)"
+    end
+
+    def api_path
+      path = definition.path
+      path = path.gsub("/repos/:owner/:repo", "\#{Repository.path repo}")
+      path = definition.params.select(&:required).reduce(path) do |path, param|
+        path.gsub(":#{param.name}", "\#{#{param.name}}")
+      end
+    end
+
+    def return_type_description
+      if definition.verb == "GET" && !resource.singular?
+        "[Array<Sawyer::Resource>]"
+      else
+        "<Sawyer::Resource>"
+      end
+    end
+
+    def parameter_type(parameter)
+      {
+        "repo" => "[Integer, String, Repository, Hash]",
+      }[parameter.name] || "[#{parameter.type.capitalize}]"
+    end
+
+    def parameter_description(parameter)
+      return parameter.description unless parameter.description.empty?
+      return "A GitHub repository" if parameter.name == "repo"
+
+      "The ID of the #{parameter.name.gsub("_id", "").gsub("_", " ")}" if parameter.name.end_with?("_id")
+    end
+
+    def parameter_documentation
+      definition.params.select(&:required).map {|param|
+        "@param #{param.name} #{parameter_type(param)} #{parameter_description(param)}"
+      } + definition.params.reject(&:required).reject {|param| ["per_page", "page"].include?(param.name)}.map {|param|
+        "@param options [#{param.type.capitalize}] :#{param.name} #{param.description}"
+      }
+    end
+
+    def return_value_description
+      case verb
+      when "GET"
+        if resource.singular?
+          "A single #{resource.name.gsub("_", " ")}"
+        else
+          "A list of #{resource.name.gsub("_", " ")}"
+        end
+      when "POST"
+        "The new #{resource.name.singularize.gsub("_", " ")}"
+      else
+      end
+    end
+
+    def verb
+      definition["method"]
+    end
+
+    def parameters
+      parameterizer.parameterize(arguments)
+    end
+
+    def arguments
+      definition.params.select(&:required).map(&:name)
```

Comment:
I had been programming for several years before I realized the difference between parameters (part of the method signature) and arguments (the inputs when you call that method).